grpc-testtool/benches/insert.rs

```diff
@@ -0,0 +1,235 @@
+                    Some(full_path)
+                } else {
+                    None
+                }
+            })
+            .next()
+    })
+}
+
+/// The actual insert benchmark
+fn insert<const BATCHSIZE: usize>(criterion: &mut Criterion) {
+    // We don't want to keep reconnecting to the DatabaseClient while running this
+    // benchmark, so we stash it here. We can't initialize it here because we need
+    // to do this in an async context.
+    let client = OnceCell::new();
+
+    // we save the tokio runtime because the client is only valid from within the
+    // same runtime
+    let runtime = tokio::runtime::Runtime::new().expect("tokio startup");
+
+    // clean up anything that was running before, and make sure we have an empty directory
+    // to run the tests in
+    reset_everything();
+
+    // We want a consistent seed, but we need different values for each batch, so we
+    // reseed each time we compute more data from the next seed value upwards
+    let seed = RefCell::new(0);
+
+    criterion.bench_with_input(
+        BenchmarkId::new("insert", BATCHSIZE),
+        &BATCHSIZE,
+        |b, &s| {
+            const KEYLEN: usize = 32;
+            const DATALEN: usize = 32;
+
+            b.to_async(&runtime).iter_batched(
+                || async {
+                    let client: &Result<
+                        DatabaseClient<tonic::transport::channel::Channel>,
+                        tonic::transport::Error,
+                    > = client.get_or_init(DatabaseClient::connect(TESTURI)).await;
+                    let client = client.as_ref().expect("connection succeeded");
+
+                    // seed a new random number generator to generate data
+                    // each time we call this, we increase the seed by 1
+                    // this gives us different but consistently different random data
+                    let seed = {
+                        let mut inner = seed.borrow_mut();
+                        *inner += 1;
+                        *inner
+                    };
+                    let mut rng = rand::rngs::StdRng::seed_from_u64(seed);
+
+                    // generate the put request, which is BATCHSIZE PutRequest objects with random keys/values
+                    let put_requests: Vec<PutRequest> = (0..s)
+                        .map(|_| {
+                            (
+                                rng.borrow_mut()
+                                    .sample_iter(&Alphanumeric)
+                                    .take(KEYLEN)
+                                    .collect::<Vec<u8>>(),
+                                rng.borrow_mut()
+                                    .sample_iter(&Alphanumeric)
+                                    .take(DATALEN)
+                                    .collect::<Vec<u8>>(),
+                            )
+                        })
+                        .map(|(key, value)| PutRequest { key, value })
```

Comment:
Is this (calling map twice) considered more idiomatic than just doing one call to map like this?

```
                    let put_requests: Vec<PutRequest> = (0..s)
                        .map(|_| PutRequest {
                            key: rng
                                .borrow_mut()
                                .sample_iter(&Alphanumeric)
                                .take(KEYLEN)
                                .collect::<Vec<u8>>(),
                            value: rng
                                .borrow_mut()
                                .sample_iter(&Alphanumeric)
                                .take(DATALEN)
                                .collect::<Vec<u8>>(),
                        })
                        .collect();
```